web/api/v1/api.go

```diff
@@ -482,6 +484,63 @@ func (api *API) targets(r *http.Request) (interface{}, *apiError) {
 	return res, nil
 }
 
+func (api *API) targetMetadata(r *http.Request) (interface{}, *apiError) {
+	limit := -1
+	if s := r.FormValue("limit"); s != "" {
+		var err error
+		if limit, err = strconv.Atoi(s); err != nil {
+			return nil, &apiError{errorBadData, fmt.Errorf("limit must be a  number")}
+		}
+	}
+
+	matchers, err := promql.ParseMetricSelector(r.FormValue("match"))
+	if err != nil {
+		return nil, &apiError{errorBadData, err}
+	}
+
+	var metric string
+	for i, m := range matchers {
+		if m.Name == labels.MetricName && m.Type == labels.MatchEqual {
+			metric = m.Value
+			matchers = append(matchers[:i], matchers[i+1:]...)
+			break
+		}
+	}
+	if metric == "" {
+		return nil, &apiError{errorBadData, errors.New("matcher must specify metric name")}
```

Comment:
I don't think this should be required, you should be able to dump all metadata for one target.

Reply:
Yes, was just about to add that as a "list" implementation that's still per-target.
After some consideration, a global list will generate massive amounts of data and seems to dangerous.

Reply:
Yea, this requires some thought but it'd still be very useful to get an (arbitrarily) deduped list.

Reply:
I'd opt for providing the metadata first and waiting for feature requests. Better than providing something now with guessed semantics and risky performance implications.